api/post/types.ts

```diff
@@ -0,0 +1,36 @@
+import { Person } from 'api/person/types'
+import { sanityArrayOf } from 'api/sanity/array/types'
+import { SanityBlock } from 'api/sanity/block/types'
+import { SanityDocument } from 'api/sanity/document/types'
+import { SanityReference } from 'api/sanity/reference/types'
+import { z } from 'zod'
+
+/**
+ * Exhaustive definition
+ */
+const PostAttributes = z.object({
+  title: z.string(),
+  author: SanityReference,
+  body: z.array(sanityArrayOf(SanityBlock)),
+})
+
+const Post = SanityDocument.merge(PostAttributes)
+
+/**
+ * Query definitions
+ */
+export const postProjection = `
+  {
+    _id,
+    title,
+    body,
+    author->
+  }
+`
+export const postsQuery = `*[_type == "post"]${postProjection}`
+
+const PostPick = Post.pick({ _id: true, title: true, body: true })
+
+const PostRes = PostPick.merge(z.object({ author: Person }))
+export const Posts = z.array(PostRes)
+export type TPost = z.infer<typeof PostRes>
```

Comment:
Reeeeally wanted to colocate the query and the type data of the expected query response, despite `postsQuery` not really fitting the `/types.ts` filename.

We might find that moving to/building a typed query builder would yield an easier-to-grok interface, but I'd prefer to wait for that to become obvious before introducing that complexity.